### DISCUSSION TITLE
RNA types: tRNA

### DISCUSSION THREAD

**Discussion by eboileau** [dieterich-lab/scimodom#69] [Features]
We started adding data, now only mRNA. The display is based on [bedRMod](https://dieterich-lab.github.io/scimodom/bedrmod.html), and shows site-specific modifications arranged by _chrom_, _start_, _strand_, and includes, among other fields,  _score_, _coverage_, and _frequency_ (dataset-specific information).

[Ensembl](https://www.ensembl.org/index.html) annotation has all tRNA genes, and we are using Ensembl already, so this is the default choice. When searching, we can filter by RNA type to display either tRNA, mRNA, *etc.*, *using the same format.*

**In general, how do you annotate your tRNA data? What about other RNA types annotation (mt-tRNA, ...)?**

**Comment by inaarmann** [comment]
Hi @eboileau ,

tRNA modifications are usually annotated according to the clover leaf structure, see e.g. 
https://www.nature.com/articles/s41580-021-00342-0

This is usually not consistent with the Ensembl coordinates.

@piechottam has done quite some work here to deal with this.

For tRNAs, a column with the "clover leaf position" would be helpful.

**Comment by piechottam** [comment]
tRNAs can  be classified into:

* isodecoder: same anticodon and amino acid but "different sequence"
* isoacceptor: different anticodon but the same amino acid

This is reflected in their naming: http://gtrnadb.ucsc.edu/docs/naming/
A tRNA gene symbol consists of 5 components:
prefix:amino-acid:anti-codon:transcript_id:gene_locus.
I have been only working with annotation from gtrnadb.ucsc.edu and modomics. I don't know how well they interact with ENSEMBL.

Biologists preferably refer to positions within tRNA by cloverleaf or Sprinzl coordinates
(see https://www.ncbi.nlm.nih.gov/pmc/articles/PMC147216/).
It is a coordinate system based on a consensus secondary structure alignment (cloverleaf) of distinct tRNA sequences - a specific set for Eukaryotes, Bacteria, and Archaea.

Some tRNAs lack some coordinates, e.g., [...], 16, 18, [...] (position 17 of the consensus model is missing). 
Then, other tRNAs have additional positions: [...], 16, 17, 17A, 18, [...]. 
Some other tRNA contain whole stretches (variable arm): e11, 12, e1, e2, e3, e4, e5, e22, e21. With a particular naming scheme where the naming reflects the secondary structure interaction: (= corresponds to base pairing) e11 = e21 and e12= e22. The e1 - e5 correspond to the loop with no base pairing.
Additionally, some tRNA have introns that get spliced during maturation. Those positions are denoted by i(1-23).

To transform genomic to Sprinzl coordinates, you require an appropriate consensus secondary structure model (correct organism, Introns (yes/no) ... and you need the sequences (tRNAs) that you want to transform. As a result, you get a mapping file with sequence coordinates -> Sprinzl coordinates. I have done this in QutRNA and can provide you the mapping files. 

My thoughts:
* I think Sprinzl coordinates are essential and should be added as "additional" data.
* For Feature, you could use: Exon, Intron, and variable-arm (or trna-variable-arm to make it clear it is tRNA specific)
* For Biotype: tRNA
* Gene: gtrnadb naming

**Comment by eboileau** [comment]
Thanks @piechottam 

Our starting point is a bedRMod file *i.e.* BED9+2 file, with single sites, so

- if this file is in genomic coordinates, there is no simple solution, as for mapping from genomic to Sprinz _you need the sequences (tRNAs) that you want to transform_. The only DB that I know of that does this sort of conversion is [RMBase](https://rna.sysu.edu.cn/rmbase3/index.php), but the concept is different.

- if this file is already in Sprinzl coordinates, then I assume we could use chrom = gtrnadb naming, start/end = Sprinzl coord/Sprinzl coord+1, but what is strand? Then we also have the problem that we don't really know which consensus it was generated from, whether it includes or not some of the coordinates, variable arms, introns, *i.e.* for genomic coordinates, position 15 on chrom 1 is the same for everyone, but in this case, how do we know if position 15 in Sprinzl coordinates is consistent across different uploaded dataset?

While your reply really helps to put this in perspective, I don't really know what to do.

**Reply by piechottam** [reply]
As far as I understand, you have a bedRMod file and two options for coordinates:
* genomic coordinates
* Sprinzl

I don't know the bedRmod file format, but I would expect it to have an assembly version key. This would also include organism info and could be used for genomic and Sprinzl coordinates. The consensus is irrelevant, assuming that the underlying genome and annotation are known. Assuming that the Sprinzl coordinates are correct and given the genome annotation, you can use the gtrnadb gene symbol to calculate genomic coordinates for the tRNA. Given the annotation from gtrnadb, we can identify the strand as well.


I could create mapping files for genomic coordinates. Given the genomic coordinates, one can identify the tRNA and apply the abovementioned transformation. The mapping files have to be created only once per assembly. (I have attached an example for Mouse - the content should be clear from the column names [mapping.txt](https://github.com/user-attachments/files/15566517/mapping.txt)). This allows Sprinzl <-> genome.



My problem is that I would keep modifications on tRNA only on genomic coordinates and add Sprinzl coordinates as extra columns. I would expect bedtools to fail when encountering positions such as: 17a, 20a, or e12. 
Sprinzl coordinates do not have a strand. One could always set it to "+" on the "coding" strand.

**Comment by eboileau** [comment]
bedRMod has the following columns:

    Chromosome (only chroms, we exclude contigs, scaffolds in SCIMODOM)
    Start
    End - typically Start + 1 for single site modification
    Name of modification (Modomics short name)
    Score
    Strand
    Thick start - generally unused
    Thick end - generally unused
    Color value (RGB)
    Coverage, or number of reads
    Frequency or percentage of modified reads at this position

Yes, we know the organism, and the assembly is fixed in SCIMODOM, *e.g.* we have GRCm39 for mouse, and if a dataset is GRCm38, then data records are lifted over during upload. 

So, if I understand correctly, we could *in principle* allow a user to decide which coordinate system he wants to use for it's tRNA bedRMod file? 

- If it's genomic coordinates *e.g.* `chr1	34434811	34434812 ...` then we lift over, if need be, then using gtrnadb annotation for the current assembly, we find the gtrnadb id or gene name `tRNA-Glu-TTC-2-1`, then using your mapping, we add "a Sprinzl column". 

- If it's in Sprinzl coordinates, then it has to be for the current assembly, as we don't want to keep mappings and annotations for different assemblies. In that case, bedRMod chrom has to be the gtrnadb id or gene name, and start/end the Sprinzl coordinates. Using your mapping, we find the sequence position, then using the gtrnadb annotation, we find the genomic coordinates.

@piechottam let me know if I didn't understand correctly. I still have to wrap my head around the mapping though, I need to try it...
Where do you define the "codes" for `ss`, and how do you "calculate" `aln_pos`? I also see that you keep introns in there, if present. I guess we could retrieve the features (exon, intron, or variable-arm) from the mapping. 

As for the gtrnadb annotation, I found some BED12 files, but I'm not sure whether you suggest using a particular file? I don't find an easy up-to-date API to download files, any suggestion?

**Comment by piechottam** [comment]
In principle, you could allow genomic and Sprinzl coordinates.
The problem is the specification of BED files: start and stop do not support letters to my knowledge, e.g., 17a, 20a e1, e12, etc. (Sprinzl coordinates). I would assume that most of the tools that process BED-files, for example, bedtools, would through an error.

My suggestion: don't implement Sprinzl coordinates for columns: Start, End. Add an extra column for metadata, e.g.: sprinzl=17A.

You are correct about the mapping: genomic -> Sprinzl and Sprinzl -> genomic. However, as mentioned earlier, I would not support Sprinzl coordinates via columns Start and End.

The mapping to be done ONCE for an assembly(organism) and release of gtrnadb.

Assuming, a reference  FASTA (genomic coordinates) for tRNAs for some organisms is given.
Pick the correct consensus secondary structure alignment 
(-> for human use euct, for mitochondria use bacterium ...
Here are some links to models that I use:
https://github.com/UCSC-LoweLab/tRAX/blob/master/TRNAinf-euct.cm
https://github.com/UCSC-LoweLab/tRAX/blob/master/TRNAinf-bact.cm
https://github.com/UCSC-LoweLab/tRAX/blob/master/TRNAinf-arch.cm

In the next step, you use cmalign to calculate a secondary structure alignment with your FASTA against a CM model. 
If you parse the resulting alignment, you can translate sequence coordinates from FASTA to secondary structure elements (e.g., "(")) and Sprinzl coordinates against the consensus model.

The scripts that I have written for this are:

https://github.com/dieterich-lab/QutRNA/blob/main/workflow/scripts/ss_consensus_to_sprinzl.py
This will produce a file with Sprinzl coordinates in the consensus alignment (of FASTA sequences) against the CM model. 

https://github.com/dieterich-lab/QutRNA/blob/main/workflow/scripts/seq_to_sprinzl.py
This parses each alignment and assigns Sprinzl coordinates from the previous script.

**Comment by eboileau** [comment]
bedtool wouldn't be a problem as such, as operations would only be done on genomic coordinates, but indeed that would be "against" the specs to allow start/end to be a string. Adding an extra column is a possibility, but in our case we need to be careful as this would make exactly 12 columns, and some tools, like bedtools, then assume it's a BED12 file. But we can handle this internally.

**Comment by eboileau** [comment]
Some questions about mapping genomic <-> Sprinzl

1. _e.g._ some sites may be "unmappable", do you see this? In general the GtRNAdb FASTA contains less sequences than what is annotated (BED), so even if a site can be annotated in genomic coordinates, we may not be able to find it's Sprinzl position; or if starting with Sprinzl position, we may fail to find the genomic coordinate to annotate it.

2. Do you see sites that cannot be annotated? _e.g._ someone may have used a custom annotation, while we're using GtRNAdb 21. _cf._ for standard RNA-seq, everything outside is intergenic, but for tRNA? Even redefining "intergenic" accordingly, does it make sense to call such sites intergenic?

**Reply by CDieterich** [reply]
@eboileau - please let me knwo if I can be of any help.
Generally, tRNA loci on the genome are known.. so there should be a 1:1 correspondence to the FASTA file.

**Comment by eboileau** [comment]
So what happens is 

1. GtRNAdb sequences are for the high confidence set, while annotations contains more, but we can restrict annotations to match the sequences used to construct the mapping, so the problem reduces to

2. what if a dataset contains sites that are not annotated (typically not mapped to high confidence loci), e.g. tRX. Either we keep them w/o annotation, or discard them.
